pkg/net/option.go

```diff
@@ -62,12 +62,32 @@ func WithRand(r *rand.Rand) Option {
 
 		r = globalRand
 	}
+	ar := &atomicRand{r: r}
 
 	return func(o *Overlay) {
-		o.n.vtx.r = r
+		o.r = ar
+		o.n.vtx.r = ar
 	}
 }
 
+type atomicRand struct {
+	r  *rand.Rand
+	mu sync.Mutex
+}
+
+func (ar *atomicRand) Intn(n int) int {
+	ar.mu.Lock()
+	result := ar.r.Intn(n)
+	ar.mu.Unlock()
```

Comment:
Let's rewrite this using `defer`.  Since we're returning a value, it ends up being more readable.

```go
ar.mu.Lock()
defer ar.mu.Unlock()

return ar.r.Intn(n)
```

Reply:
I also thought about this option, but it felt more readable the other way. Anyway, I will implement your suggestion, because I'm not strongly opinionated on this.